File: algorithms/lru.cpp
/*
 * @Author: your name
 * @Date: 2020-12-25 20:47:29
 * @LastEditTime: 2020-12-25 22:30:44
 * @LastEditors: Please set LastEditors
 * @Description: In User Settings Edit
 * @FilePath: /ivan个人练习代码/algorithms/lru.cpp
 */
#include <iostream>
using namespace std;
#define AP 3
int total_instruction = 0;
int diseffect = 0;
int page[AP];
int pagecontrol[100];

void init(){
   for(int i = 0; i < AP;i++){
       page[i] = -1;
   }
   cout<<"输入要访问页面访问:"<<endl;


while(cin.peek()!= '\n'){
    cin >> pagecontrol[total_instruction];
    total_instruction += 1;
}
}
void show_page(int t){
    cout<< t << ": ";
    for(int i = 0; i <AP;i++){
        cout<<page[i]<<" ";

    }
    cout<<endl;
}
void LRU(){
   
   for(int i = 0;i<total_instruction;i++){
       int flag = 0;
       int last_page;
       for(int j = 0;j<AP;j++){
           if (page[j] == pagecontrol[i])
           {
            last_page = page[i];
            flag = 1;
            for(int k = j;k<AP - 1;k++)
            {
                page[AP - 1] = last_page;
                break;
            }


           }
           if(flag == 0){
               diseffect++;
               for(int k = 0; k < AP - 1; k++){
                   page[k] = page[k+1];
               }
               page[AP - 1] = pagecontrol[i];
           }
           show_page(i);
           
       }
   }
     
}
 int main(){
          init();
          LRU();
          cout<< "总缺页次数: "<<diseffect<<endl;
          cout<< "缺页率: "<<diseffect*1.0 / total_instruction * 100 <<"%"<<endl;
      }

File: algorithms/lru.js
/*
 * @Author: your name
 * @Date: 2021-03-15 15:44:02
 * @LastEditTime: 2021-03-15 15:44:37
 * @LastEditors: Please set LastEditors
 * @Description: In User Settings Edit
 * @FilePath: /ivan个人练习代码/algorithms/lru.js
 */
/**
 * @param {number} capacity
 */
var LRUCache = function(capacity) {
    this.map = new Map()
    this.capacity = capacity
};

/** 
* @param {number} key
* @return {number}
*/
LRUCache.prototype.get = function(key) {
   var val = this.map.get(key)
   if(val == undefined) return -1
   this.map.delete(key)
   this.map.set(key,val)
   return val
};

/** 
* @param {number} key 
* @param {number} value
* @return {void}
*/
LRUCache.prototype.put = function(key, value) {
   if(this.map.has(key)) this.map.delete(key)
   this.map.set(key,value)
   if(this.map.size >  this.capacity){
       this.map.delete(this.map.entries().next().value[0])
   }
};




/**
* Your LRUCache object will be instantiated and called as such:
* var obj = new LRUCache(capacity)
* var param_1 = obj.get(key)
* obj.put(key,value)
*/